### PR TITLE
Add Infrastructure as Code template for function deployment

### PR DESCRIPTION
**Pull request by nmiodice**
This change adds an IaC deployment template to provision the resources needed to host this application, as well as all of its dependencies. Among the provisioned resources are:

* Storage account and container used to hold [Terraform state](https://www.terraform.io/docs/state/remote.html). These are kept in a separate resource group from the service deployment.
* Resource group for `gitinsights` function deployment
* KeyVault for hosting Azure DevOps token. The token is automatically created as a secret in this vault
* Consumption App Service Plan
* Azure Function with Managed Identity that can consume secrets from KeyVault

### CODE REVIEW

**Comment by erikschlegel** [comment]
@nmiodice can you link this PR to issue #10

**Review by erikschlegel — APPROVED** [review]
This looks great. This is a much needed feature that will enable developers to provision the gitinsight Azure resources in a predictable and repeatable fashion. 👍 

Left some minor comments / questions below.

File: terraform/git-insights/secrets.tf [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  certificate_permissions = ["create", "delete", "get", "list"]
+}
+
+resource "azurerm_key_vault_access_policy" "function-get" {
+  key_vault_id       = azurerm_key_vault.kv.id
+  tenant_id          = data.azurerm_client_config.current.tenant_id
+  object_id          = azurerm_function_app.func.identity.0.principal_id
+  secret_permissions = ["get"]
+}
+
```
[the remark this answers is not in the document]
Comment (nmiodice) [reply terraform/git-insights/secrets.tf:28]: None that I know of. All we really need is to set up an access policy for the Function to pull from KV. I didn't need to do anything else in order to get the function to invoke successfully

File: terraform/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+```bash
+DOT_ENV=.env
+export $(cat $DOT_ENV | grep -v '^\s*#' | xargs)
+```
+
+## Initial Deployment
+
+### Setup Terraform Backend State
+
```
[the remark this answers is not in the document]
Comment (nmiodice) [reply terraform/README.md:23]: They are stated in the `.env.template` files in each directory, but I'll make it more clear

File: terraform/git-insights/secrets.tf [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+resource "azurerm_key_vault_secret" "token" {
+  name         = "token"
```
[the remark this answers is not in the document]
Comment (nmiodice) [reply terraform/git-insights/secrets.tf:30]: `gitinsights_pat` is an invalid secret name, but I will use `gitinsights-pat`

**Pull request merged**